man/man1/yagle.1
.TH YAGLE 1 "30 March 2000" "AVERTEC" "AVERTEC CAD Tools"

.SH NAME
yagle \- Disassembly and functional abstraction of CMOS circuits

.so man1/avt_origin.1

.SH SYNOPSIS
.B "yagle "
[
.B \-i
]
[
.B \-v
]
[
.B \-p=n
]
[
.B \-nc
]
[
.B \-nl
]
[
.B \-fcl
]
[
.B \-elp
]
[
.B \-o
]
[
.B \-z
]
[
.B \-s
]
[
.B \-nb
]
[
.B \-h
]
[
.B \-d
]
[
.B \-t
]
[
.B \-a
]
.I input_name
[
.I output_name
]

.SH DESCRIPTION
.B yagle
is a circuit disassembler and functional abstractor for CMOS digital circuits.
It generates a VHDL data flow description and an oriented gate net-list
from a transistor level description of the circuit. The transistor net-list
can be either flat or hierarchical.
.LP
The VHDL subset generated by
.B yagle
is supported by most simulation and synthesis tools.
.LP
.B yagle
does not use a predefined gate library except for latches and flip-flops.
All styles or circuitry are supported: dual-cmos, precharge, pass-transistor, etc.
In addition a user-defined gate library can be provided in order to handle complex latches
or analog circuitry.
.LP
\fBAll power supplies and grounds signals must be connected to an external connector.\fB
.LP
In the first phase,
.B yagle
extracts the CMOS dual circuitry. 
In the second phase,
.B yagle
builds the gate net-list for the remaining circuitry whilst performing functional
analysis in parallel, in order to prevent the fabrication of false branches
within a gate and to verify the behaviour of the gate.
.LP
This functional analysis depends on the '-p=n' option which defines the maximum
depth (in gates) of the analysis.
.LP
.B yagle
reads the transistor net-list given by
.I input_name
and generates a VHDL data flow description in
.I output_name.
If no
.I output_name
is given then
.I input_name
is used.

.SH OPTIONS
.sp 1
Options may be given in any order before or after the filename(s).
.LP
.TP 0.5i
.B \-v
When this option is set, the interface and the internal signal of the behavioural
description are vectorised. Every bit of the vector has to be of the same type
otherwise the functional description will not be generated. 
.TP 0.5i
.BI \-p= n
This option sets the maximum depth for the functional analysis. 
This is the depth of circuitry (in gates) taken into account when detecting
reconvergences in the circuit. The default value is 5. 
When
.I n
= 0, the functional analysis process is disabled.
.TP 0.5i
.B \-nc
Disables the detection of complex gates. Without this option complex gates such
as edge-triggered flip-flops are identified by pattern-matching applied to the
disassembled gate net-list. Special predefined behavioural descriptions are then
generated for these gates. This option is useful if a one to one correspondance
between the elements of the behavioural description and the elements of the
disassembled gate net-list is required.
.TP 0.5i
.B \-nl
Disables the detection of latches and memory points using the built-in latch
library. This option is useful if all memory points are to be detected by
the use of a user-defined library with the
.B \-fcl
option.
.TP 0.5i
.B \-fcl 
This option makes
.B yagle
use library-based transistor netlist recognition (see man fcl). This
allows the user to specify a number of netlists to be identified within the
circuit to be disassembled. These netlists are specified in the Spice format and
can contain a number of special directives for the marking of the identified
signals and transistors in the circuit, for example signals corresponding to
memory points or transistors to be ignored.
.br
In addition the user can specify a behavioural description for the transistor
netlist which is used to generate the global
behavioural description of the circuit. This allows the functional abstraction
of circuits containing analog blocks for example RAMs containing sense amplifiers.
.TP 0.5i
.B \-xfcl
Same as \-fcl. In addition, it makes yagle stop after the recognition process.
.TP 0.5i
.B \-elp
With this option set
.B yagle
uses the elp tecnology file (man elp) to update the capacitances in the input
transistor net-list. So that the cns and structural views contain accurate
capacitance values.
.TP 0.5i
.B \-o
.B yagle
orients transistors using a simple rule:
A transistor whose source is connected to the output of a CMOS Duals gate, and
not connected to a transistor gate, is oriented form source to drain. This
orientation is performed during the phase of extraction of CMOS duals.
.TP 0.5i
.B \-z
When this option is set,
.B yagle
exploits high impedance nodes during the phase of
functional analysis. This allows, for example, the resolution of false conflicts
in circuits which use precharge logic.
.TP 0.5i
.B \-s
This option provides only one power supply and ground connector in the interface
of the behavioural description. This can be useful in order to compare the 
abstracted description with the initial specification. When this option is set,
the name of power supply and ground
are given by
.B MBK_VDD
and
.B MBK_VSS.
.B \-nb
Disables the generation of the behavioural description of the circuit, useful if
the user is solely interested in the gate-level net-list.
.TP 0.5i
.B \-h
.B yagle generates a structural gate level net-list. The variable MBK_OUT_LO has to be set to
choose the format of the out files: the net-list and the gates. A behavioural
description is generated for each gate.
.TP 0.5i
.B \-d
.B yagle
generates a .cns file which contains the flat gate net-list. This file is mainly
used for debugging.
.TP 0.5i
.B \-i 
This option makes
.B yagle
read the .inf file. This file should have the same name
as the input file with extesion .inf. It may contain mutual exclusion
conditions on ports of the circuit. These conditions are only used in the
functional analysis process. 
.br
# lines beginning with '#' are comment lines
.br
Syntax for the mutual exclusion conditions
.br
	MUTEX
.br
	    muxUP{a,...,d};
.br
	    muxDN{m,...,p};
.br
	    cmpUP{i,...,l};
.br
	    cmpDN{x,...,z};
.br
	END
.br

muxUP expresses that one port at most in the list is "one".
.br
muxDN expresses that one port at most in the list is "zero".
.br
cmpUP expresses that one and only one port in the list is "one".
.br
cmpDN expresses that one and only one port in the list is "zero".

.br
Port name may be preceded by the character '~' which minds that it is the inverse
of the port which has to be taken into account. 
.br

The user can also use this file to rename internal signals in the behavioural 
description in order to use the formal proof.

.br
Syntax to rename signals for the behavioural description
.br
	RENAME 
.br
	    existing_name : new_name  ;
.br
	    *gno* : *latch_data* ;
.br
	END
.br

"new_name" will replace "existing_name" in the behavioural description file. 
It is possible to  use the joker '*'. When the names contain the string "gno",
this string is replaced by the string "latch_data" (l2_y_gno_01 is replaced by
l2_y_latch_data_01).

Beware that only one rule can be applied to a name, (the following rules 
are then ignored when one has been applied) and that the rules are 
taken into account in the order in which they appear in the 'inf' file.
.TP 0.5i
.B \-t
Set the trace mode during execution, used only for debugging purposes.
.br

.TP 0.5i
.B \-a
Activate transistor orientation taking into account the  \fB_s\fP
convention on signal's names.
.br

.SH OUTPUT FILES
.TP 10
.B vbe
The functional description is described in a file called
.I input_name
or
.I output_name, with extension .vbe.
.TP 10
.B cns
The disassembled gate net-list or cone net-list description when the
.B \-d
option is set.
.TP 10
.B vst/hns/al/spi
The structural description when the
.B \-h
option is set.
.TP 10
.B rep
The errors and warnings report file.
.B loop
File containing list of combinatorial loops file, only created if loops exist
and the environment variable
.B YAGLE_LOOP_MODE
is set.

.SH ENVIRONMENT VARIABLES
.TP 7
MBK_IN_LO
indicates the format of the input net-list.
.br

	spi	for	Spice net-list.                
.br
	al	for	Alliance extracted net-list.
.br

.TP 7
.B MBK_OUT_LO
indicates the format of the output net-list when the 
.B \-h
option is set. Same values as
.B MBK_IN_LO
.TP 7
.B MBK_WORK_LIB
Indicates where YAGLE has to read the input file and write the resulting files.
.TP 7
.B MBK_CATA_LIB
If the input net-list is hierarchical, the leaf cells may not be in the working
directory MBK_WORK_LIB. In that case, MBK_CATA_LIB indicates where YAGLE can
find the cells.
.TP 7
.B CNS_VDDNAME 
Sets the name of power supply. "vdd" is the default. Every external port of the
circuit whose name contain this string will be considered as a power supply.
.TP 7
.B CNS_VSSNAME
Sets the name of the ground. "vss" is the default.  Every external port of the
circuit whose name contain this string will be considered as a ground.
.TP 7
.B CNS_GRIDNAME
Sets the name of the grid connector of a transistor. "grid" is the default.
.TP 7
.B CNS_SOURCENAME
Sets the name of the source connector of a transistor. "source" is the default.
.TP 7
.B CNS_DRAINNAME
Sets the name of the drain connector of a transistor. "drain" is the default.
.TP 7
.B ELP_TECHNO_NAME
Sets the full access path and name of the technology file (.elp) used to correct
the node capacitances in the circuit.
.TP 7
.B VH_BEHSFX
Sets the extension of the file that will contain the VHDL description. "vbe" is
the default value.
.TP 7
.B YAGLE_LANGUAGE
When set to F,
.B yagle
will report errors and warnings in French.
When set to E,
.B yagle
will report errors and warnings in English.
The default is English.
.TP 7
.B YAGLE_STAT_MODE
When set to Y,
.B yagle
generates a file with the extension .stat which contains statistics of the
transistor net-list.
.TP 7
.B YAGLE_MAX_LINKS
Sets the the maximum possible length of a branch within a gate. The default is
six.
.TP 7
.B YAGLE_LOOP_MODE
When set to Y,
.B yagle
looks for combinatorial loops in the disassembled circuit, if any are found,
they
are reported in a file with extension
.B \.loop.
Note that any two gate loops are systematically reported in the
.B \.rep
file.
.TP 7
.B YAGLE_BDDCEILING
Sets the the maximum number of BDD nodes. The default is 10000, 0 means no ceiling.
.TP 7
.B YAGLE_BLOCK_BIDIR
When set to 'yes', this variable inhibits the construction of branches through a transistor already used
in the opposite orientation.
.TP 7
.B YAGLE_GEN_SIGNATURE
When set to 'no', yagle does not generate signatures for the cones, therefore .slib icones are not
used.
.TP 7
.B FCL_LIB_PATH
Indicates the access path to the directory containing the user-defined cell
library used if the
.B \-fcl
option is set. The default is a subdirectory
.B /cells
in
.B MBK_WORK_LIB
.TP 7
.B FCL_LIB_NAME
The name of the file (located in
.B FCL_LIB_PATH
) containing the list of cells in the user-defined cell library used if the
.B \-fcl
option is set. The default is
.B LIBRARY.
.SH DIAGNOSTICS
Reported in the
.B rep
file.
.br
"[WAR] Possible unconnected supply ?"
.br
means that an internal signal whose name contains
.B CNS_VDDNAME
or
.B CNS_VSSNAME
has been found. Verify if this signal should be connected to an external supply.
.br
"[WAR] Transistor used as a resistance" 
.br
Indicates that a transistor P (resp. N) with gate connected to the ground
(resp. power supply) has been found in the circuit.
.br
"[WAR] Transistor used as a diode" 
.br
Indicates that a transistor with drain (or source) connected to gate has been
found in the circuit, and the signal connecting them is neither power supply nor
ground.
.br
"[WAR] Transistor is always off" 
.br
Indicates that a transistor P (resp. N) with gate connected to power supply
(resp. ground) has been found in the circuit.
.br
"[WAR] Transistor used as a capacitance" 
.br
Indicates that a transistor with drain and source connected together has been
found in the circuit.
.br
"[WAR] Gate of transistor is not connected" 
.br
Indicates that a transistor gate which is connected to nothing has been found
in the circuit.
.br
"[WAR] Drain of transistor is not connected" 
.br
Indicates that a transistor drain which is connected to nothing has been found
in the circuit.
.br
"[WAR] Source of transistor is not connected" 
.br
Indicates that a transistor source which is connected to nothing has been found
in the circuit.
.br
"[WAR] Transistors are not used in the circuit" 
.br
This means that these transistor are not used to pull up or pull down any
transistor gate in the circuit, or any external port. This occurs if the output
of an inverter does not drive anything: In this case
.B yagle
considers both
transistors of the inverter to be unused.
.br
"[WAR] Loop between 2 gates (bleeder found)" 
.br
This means that a loop corresponding to a bleeder has been found in the circuit. 
.br
"[WAR] Loop between 2 gates (latch found)" 
.br
This means that a loop corresponding to a latch has been found in the circuit. 
.br
"[WAR] Loop between 2 gates (bistable found)" 
.br
This means that a loop corresponding to a bistable has been found in the circuit. 
.br
"[WAR] Loop between 2 gates (nothing found)" 
.br
This means that a two gate loop which does not correspond to a latch, bleeder
or bistable has been found in the circuit. 
.br
"[WAR] Conflict may occur on signal" 
.br
This means that the signal may be pulled-up and pulled-down simultaneously. This
is a warning since this message  may disappear with a greater depth for the
functional analysis process. 
.br
"[WAR] HZ state may occur on signal" 
.br
This means that the signal is not pulled up or pulled down for every input
pattern on the cone entries.  This is a warning since this message may disappear
with a greater depth for the functional analysis process.
.br
"[WAR] Signal does not drive anything"
.br
This means that the the signal is not used as the input to any gate or used
to drive any external connector.
.br
"[WAR] Connector unused" 
.br
This means that the external connector is neither the input nor the output of
any of the extracted transistor gates.
.br
"[ERR] Bad direction on connector" 
.br
Indicates that the orientation of an external connector after disassembly does not correspond 
to that specified in the input net-list.
.br
"[ERR] Transistor gate signal is not driven" 
.br
Indicates that a transistor gate can not be pulled up or down.
.SH FATAL ERRORS
"[FATAL ERR] No VDD/VSS connector in the circuit" 
.br
This means that
.B yagle
did not find any external ports whose name is the name of the power supply in the
circuit. Have CNS_VDDNAME and CNS_VSSNAME the right value? 
.br
"[FATAL ERR] Connector is power supply and ground" 
.br
This means that
.B yagle
found a connector whose name includes
.B CNS_VDDNAME
and
.B CNS_VSSNAME.
"[FATAL ERR] No VDD/VSS signal in the circuit" 
.br
This means that
.B yagle
did not find any signal whose name is the name of the power supply in the
circuit.
.br
"[FATAL ERR] Several external connectors on signal"
.br
This means that
.B yagle
found several external connectors connected to the same equipotential, a configuration
which it considers illegal.
.br

.so man1/avt_bug_rprt.1

.SH SEE ALSO
.BR inf (5), 
.BR fcl (5), 


debian/changelog
tasyag (3.4p5-1) unstable; urgency=low

  * Initial release. Closes: #00001

 -- Jean-Paul Chaput <Jean-Paul.Chaput@lip6.fr>  Tue, 15 May 2011 17:30:00 +0100


sources/yagle/yagle/yagle.c
/****************************************************************************/
/*                                                                          */
/*                      Chaine de CAO & VLSI   AVERTEC                      */
/*                                                                          */
/*    Produit : YAGLE  v3.50                                                */
/*    Fichier : yag_main.c                                                  */
/*                                                                          */
/*    (c) copyright 2001                                                    */
/*    Tous droits reserves                                                  */
/*                                                                          */
/*    Auteur(s) : Anthony LESTER                        le : ../../....     */
/*                                                                          */
/*    Modifie par :                                     le : ../../....     */
/*    Modifie par :                                     le : ../../....     */
/*    Modifie par :                                     le : ../../....     */
/*                                                                          */
/****************************************************************************/

#include "yag_headers.h"
#include "yag_timing.h"

#ifndef WITHOUT_TAS
#include MCC_H
#include LIB_H
#include TLF_H
#include STM_H
#include TRC_H
#endif


/* Static function declarations */

static short yagGetArgs __P((int argc, char *argv[]));
static char *optionIdent __P((char *name, char *chaine));

/****************************************************************************
*                         fonction main();                                 *
****************************************************************************/
int main(argc, argv)
    int     argc;
    char    *argv[];
{
    char error[YAGBUFSIZE];
    
#ifdef AVERTEC
    avtenv();
#endif

    mbkenv();               /* MBK environment */

    cnsenv();               /* CNS environment */
    yagenv(NULL);           /* YAG environment */
    fclenv();               /* FCL environment */
    rcnenv();
    rcx_env();
#ifdef AVERTEC_LICENSE
    if(avt_givetoken("YAGLE_LICENSE_SERVER", YAG_CONTEXT->YAGLE_TOOLNAME)!=AVT_VALID_TOKEN) EXIT(1);
#endif

/*----------------------------------------------------------------------------*
|      OPTIONS                                                                |
+-----------------------------------------------------------------------------*/

    if(yagGetArgs(argc, argv) == -1) {
        printf("\nUsage   : yagle [options] <file1> [file2]\n");
        printf("Options : -i        reads the '.inf' file\n");
        printf("          -fig=str  str is the name of the figure to handle\n");
        printf("          -v        vectorises the vhdl description\n");
        printf("          -hier     hierarchical disassembly mode\n");
        printf("          -p=n      n is the depth for functional analysis\n");
        printf("          -nc       no detection of complex gates (cells)\n");
        printf("          -cs       allow cells to share gates\n");
        printf("          -ffs      mark slave instead of master in a flip-flop\n");
        printf("          -nl       no latch recognition\n");
        printf("          -la       automatic latch detection\n");
        printf("          -ba       automatic bus analysis\n");
        printf("          -fcl      transistor netlist recognition\n");
        printf("          -xfcl     only transistor netlist recognition\n");
        printf("          -gns      hierarchical recognition\n");
        printf("          -xg       only hierarchical recognition\n");
        printf("          -tdmax    behavior with max delays\n");
        printf("          -tdmin    behavior with min delays\n");
        printf("          -tdmed    behavior with average delays\n");
        printf("          -elp      use the technology file '.elp'\n");
        printf("          -o        transistor orientation\n");
        printf("          -a        to take the '_s' convention into account\n");
        printf("          -z        functional analysis through HZ nodes\n");
        printf("          -s        only one vdd and vss in the vhdl description\n");
        printf("          -nb       no vhdl file generated\n");
        printf("          -ch       generates a hierarchical cone netlist\n");
        printf("          -d        generates a .cns file\n");
        printf("          -c        generates a .cns file with signals and RC networks\n");
        printf("          -bx       generates an intermediate netlist without black boxes\n");
        printf("          -bi       black boxes become instances\n");
        printf("          -ns       unconnected transistor gates are warnings\n");
        printf("          -nrc      suppression of RC before disassembly\n");
        printf("          -nrp      suppress parallel transistor reduction\n");
        printf("          -mi       remove simple inversions\n");
        printf("          -t        display execution trace\n");
        printf("          file2     is the vhdl file to be generated (default is file1)\n");
        printf("\n");
        yagExit(-1);
    }

/*----------------------------------------------------------------------------*
| Environment for transistor models and library preload                       |
+-----------------------------------------------------------------------------*/

#ifndef WITHOUT_TAS
    elpenv();
    mccenv();
    stmenv();
    elpLoadOnceElp();

    readlibfile("lib", lib_parse, 0);
    readlibfile("tlf3", parsetlf3, 0);
    readlibfile("tlf4", parsetlf4, 0);
#endif
    cbhenv();
    
/*----------------------------------------------------------------------------*
| Install the signal handlers                                                 |
+-----------------------------------------------------------------------------*/

    avt_TrapKill ();
    avt_PushKillExit (yagExit, 100);
    sprintf (error, "%s%s", "\n[YAGLE] A fatal internal error has occured\n",
                            "        rerun with -nb option and consult report\n");
    avt_TrapSegV ();
    avt_PushSegVMessage (error);
    avt_PushSegVExit (yagExit, 100);

    yagle_main();
    EXIT(0) ;
}

/****************************************************************************
*                         function yagGetArgs();                            *
****************************************************************************/
/*--------------------------------------------*/
/* Read the command line options              */
/*--------------------------------------------*/

static short 
yagGetArgs(argc, argv)
    int             argc;
    char           *argv[];
{
    int             i;
    short           fileExist = 0;
    char           *car;
    char           *pt;

    if (argc < 2)
    return (-1);

    for (i = 1; i < argc; i++) {
        /*----------------------------+
        | Traitement des options      |
        +----------------------------*/
        if (*(argv[i]) == '-') {

            /*--------------------------+
            | mode trace                |
            +--------------------------*/
            if (strcmp(argv[i], "-t") == 0) {
                avt_setlibloglevel (1,LOGYAG);
                continue;
            }

            /*----------------------+
            | func. analysis depth  |
            +----------------------*/
            else if ((car = optionIdent(argv[i], "-p=")) != NULL) {
                for (pt = car; *pt != '\0'; pt++)
                    if ((*pt < '0') && (*pt > '9')) return (-1);
                YAG_CONTEXT->YAG_DEPTH = atoi(car);
                if (YAG_CONTEXT->YAG_DEPTH == 0) YAG_CONTEXT->YAG_USE_FCF = FALSE;
                continue;
            }
            /*----------------------+
            | figure name           |
            +----------------------*/
            else if ((car = optionIdent(argv[i], "-fig=")) != NULL) {
                YAG_CONTEXT->YAG_FIGNAME = namealloc(car);
                continue;
            }
            /*----------------------+
            | hierarchical mode     |
            +----------------------*/
            else if (strcmp(argv[i], "-hier") == 0) {
                YAG_CONTEXT->YAG_HIERARCHICAL_MODE = TRUE;
                continue;
            }
            /*------------------------+
            | single supply in vbe    |
            +------------------------*/
            else if (strcmp(argv[i], "-s") == 0) {
                YAG_CONTEXT->YAG_ONE_SUPPLY = TRUE;
                continue;
            }
            /*----------------------+
            | inf info file         |
            +----------------------*/
            else if (strcmp(argv[i], "-i") == 0) {
                avt_sethashvar("avtReadInformationFile", inf_GetDefault_AVT_INF_Value());
                continue;
            }
            /*----------------------+
            | no behavioural file   |
            +----------------------*/
            else if (strcmp(argv[i], "-nb") == 0) {
                YAG_CONTEXT->YAG_BEFIG = FALSE;
                continue;
            }
            /*----------------------+
            | hierarchical netlist  |
            +----------------------*/
            else if (strcmp(argv[i], "-ch") == 0) {
                YAG_CONTEXT->YAG_CONE_NETLIST = TRUE;
                continue;
            }
            /*----------------------+
            | old cns netlist file  |
            +----------------------*/
            else if (strcmp(argv[i], "-d") == 0) {
                YAG_CONTEXT->YAG_FILE = TRUE;
                continue;
            }
            /*----------------------+
            | cone netlist file     |
            +----------------------*/
            else if (strcmp(argv[i], "-c") == 0) {
                YAG_CONTEXT->YAG_FILE = TRUE;
                continue;
            }
            /*----------------------+
            | orientation           |
            +----------------------*/
            else if (strcmp(argv[i], "-o") == 0) {
                YAG_CONTEXT->YAG_ORIENT = TRUE;
                continue;
            }
            /*----------------------+
            | hz node traversal     |
            +----------------------*/
            else if (strcmp(argv[i], "-z") == 0) {
                YAG_CONTEXT->YAG_PROP_HZ = TRUE;
                continue;
            }
            /*----------------------+
            | no cell detection     |
            +----------------------*/
            else if (strcmp(argv[i], "-nc") == 0) {
                YAG_CONTEXT->YAG_MAKE_CELLS = FALSE;
                continue;
            }
            /*----------------------+
            | cells can share cones |
            +----------------------*/
            else if (strcmp(argv[i], "-cs") == 0) {
                YAG_CONTEXT->YAG_CELL_SHARE = TRUE;
                continue;
            }
            /*----------------------+
            | no latch detection    |
            +----------------------*/
            else if (strcmp(argv[i], "-nl") == 0) {
                YAG_CONTEXT->YAG_DETECT_LATCHES = FALSE;
                continue;
            }
            /*----------------------+
            | loop analysis         |
            +----------------------*/
            else if (strcmp(argv[i], "-la") == 0) {
                YAG_CONTEXT->YAG_LOOP_ANALYSIS = TRUE;
                continue;
            }
            /*----------------------+
            | bus analysis          |
            +----------------------*/
            else if (strcmp(argv[i], "-ba") == 0) {
                YAG_CONTEXT->YAG_BUS_ANALYSIS = TRUE;
                continue;
            }
            /*----------------------+
            | GENIUS                |
            +----------------------*/
            else if (strcmp(argv[i], "-gns") == 0) {
                YAG_CONTEXT->YAG_GENIUS = TRUE;
                continue;
            }
            else if (strcmp(argv[i], "-xg") == 0) {
                YAG_CONTEXT->YAG_ONLY_GENIUS = TRUE;
                continue;
            }
            /*----------------------+
            | netlist detection     |
            +----------------------*/
            else if (strcmp(argv[i], "-fcl") == 0) {
                YAG_CONTEXT->YAG_FCL_DETECT = TRUE;
                continue;
            }
            else if (strcmp(argv[i], "-xfcl") == 0) {
                YAG_CONTEXT->YAG_FCL_DETECT = TRUE;
                YAG_CONTEXT->YAG_ONLY_FCL = TRUE;
                continue;
            }
            /*----------------------+
            | blackbox elimination  |
            +----------------------*/
            else if (strcmp(argv[i], "-bx") == 0) {
                YAG_CONTEXT->YAG_BLACKBOX_SUPPRESS = TRUE;
                continue;
            }
            /*----------------------+
            | blackbox ignore       |
            +----------------------*/
            else if (strcmp(argv[i], "-bi") == 0) {
                YAG_CONTEXT->YAG_BLACKBOX_IGNORE = TRUE;
                continue;
            }
            /*----------------------+
            | not strict errors     |
            +----------------------*/
            else if (strcmp(argv[i], "-ns") == 0) {
                YAG_CONTEXT->YAG_NOTSTRICT = TRUE;
                continue;
            }
            /*----------------------+
            | remove RC networks    |
            +----------------------*/
            else if (strcmp(argv[i], "-nrc") == 0) {
                YAG_CONTEXT->YAG_NORC = TRUE;
                continue;
            }
            /*----------------------+
            | remove parallel pairs |
            +----------------------*/
            else if (strcmp(argv[i], "-nrp") == 0) {
                YAG_CONTEXT->YAG_REMOVE_PARA = FALSE;
                continue;
            }
            /*----------------------+
            | remove inversions     |
            +----------------------*/
            else if (strcmp(argv[i], "-mi") == 0) {
                YAG_CONTEXT->YAG_MINIMISE_INV = TRUE;
                continue;
            }
            /*----------------------+
            | elp techno file       |
            +----------------------*/
            else if (strcmp(argv[i], "-elp") == 0) {
                YAG_CONTEXT->YAG_ELP = TRUE;
                continue;
            }
            /*----------------------+
            | tas timings           |
            +----------------------*/
            else if (strcmp(argv[i], "-tdmax") == 0) {
                YAG_CONTEXT->YAG_TAS_TIMING = YAG_MAX_TIMING;
                continue;
            }
            else if (strcmp(argv[i], "-tdmin") == 0) {
                YAG_CONTEXT->YAG_TAS_TIMING = YAG_MIN_TIMING;
                continue;
            }
            else if (strcmp(argv[i], "-tdmed") == 0) {
                YAG_CONTEXT->YAG_TAS_TIMING = YAG_MED_TIMING;
                continue;
            }
            /*----------------------+
            | use _s convention     |
            +----------------------*/
            else if (strcmp(argv[i], "-a") == 0) {
                YAG_CONTEXT->YAG_HELP_S = TRUE;
                continue;
            }
            else return (-1);
        }
        if (fileExist == 0) {
            YAG_CONTEXT->YAG_FILENAME = namealloc(argv[i]);
            YAG_CONTEXT->YAG_OUTNAME = YAG_CONTEXT->YAG_FIGNAME;
        }
        else {
            YAG_CONTEXT->YAG_OUTNAME = namealloc(argv[i]);
        }
        fileExist++;
    }

    if (YAG_CONTEXT->YAG_FIGNAME == NULL) {
        YAG_CONTEXT->YAG_FIGNAME = YAG_CONTEXT->YAG_FILENAME;
    }
    if ((fileExist == 0) || (fileExist > 2))
    return (-1);

    return (0);
}

/****************************************************************************
*                         function optionIdent();                          *
****************************************************************************/

static char *
optionIdent(name, chaine)
    char           *name;
    char           *chaine;
{
    int             i = 0;

    for (i = 0; (name[i] != '\0') && (chaine[i] != '\0'); i++)
    if (name[i] != chaine[i])
        break;

    if (chaine[i] != '\0')
    return (NULL);
    else
    return (&name[i]);
}

